Python3

oliver-sanders opened the pull request:
#### Python 2.6 -> 3.7

* For the most part a minimal effort to get Cylc running under Py3
* Dependency base is now quite pleasing :champagne:.
* All but two remote tests should pass :champagne: 
  * *batch schedulers not tested*
  * **bumping this to future work**
  * `tests/authentication/10-remote-suite-passphrase-cache.t` - awaiting conclusion of #2975 
  * `tests/cylc-message/00-ssh.t` - awaiting conclusion of #2975 
* All local tests should pass :champagne:.
* All unittests should pass :champagne:.

#### Related Issues

closes #2858 
closes #2647 
closes #2858 
closes #2864 
??? #2785 

GUI issues which become obsolete:
closes #164 
closes #1620 
closes #1671
closes #1996
closes #2466 
closes #2498 
closes #2632 
closes #2898 
closes #2633 
closes #2956 

GUI issues which we may want to keep around for the moment:
addresses #2570 
addresses #2580 
addresses #2284 


#### TODO:

* [x] Scan suites in parallel
* [x] Fix remaining test failures
* [x] Docs
* [x] pep8 (will do as last commit to make reviewing easier)
* [x] Work out what we should do with uuid (removed from client-server interface at the moment).

#### Highlights:

* Remove non-py3 compatible components
  * GTK related code
  * `cylc review` (awaiting tornado re-write)
  * cherrypy
* Upgrade isodatetime
* Upgrade to Python 3.7 (unittests are now run under 3.6 as well, functional tests should pass under 3.6, untested)
* Replace cherrypy with minimal zmq network layer
  * The aim here was to put in the absolute minimum, we will want to improve on this later
  * Anon auth has been removed awaiting a superior mechanism, tests have been disabled
* Replace `cylc graph --reference`
* Refactor `cylc scan`
  * Aggressive refactor with interface changes, I can "tone it down" a little if we think this is likely to cause problems
  * See commit notes for details

#### Further Work

* ZMQ interface improvements: #2978
* CLI improvements: #2972 
* Convert main loop to asyncio #2866 
* Cylc unicode support #2979, #2281 
* Fully replace `__cmp__` logic

**Take a look through the commits, there have been a few changes beyond Python3.**

sadielbartholomew commented:
Very happy to see this go up as a PR! :trophy: Also the line count has taken quite a satisfying ✂️ (even accounting for the fact that ``cylc review`` will return) which cannot be a bad thing; goodbye GTK & CherryPy!

oliver-sanders commented:
@kinow If you get a chance could you take a look at the two unittest failures, they are both `Mock` problems where the internals have changed so the tests have broken.

hjoliver commented:
>     * Work out what we should do with uuid (removed from client-server interface at the moment).

From a quick look, I guess you mean client UUID, not suite UUID? (the latter is needed to uniquely identify a suite run, and the new BOM infrastructure relies on this).

hjoliver commented:
>     * Scan suites in parallel

Or use asyncio?

hjoliver commented:
> [`cylc scan`] Aggressive refactor with interface changes, I can "tone it down" a little if we think this is likely to cause problems

Looks fine to me!  (from reading your commit log).

kinow commented:
Am looking at the issue with unit tests now, but wanted to give a big shout out to @oliver-sanders for his organization with commits. I know given the number of changes here, most comments will be probably regarding things that changed and maybe a few that broke (or maybe none! :crossed_fingers:). But if anyone else is curious, have a look at how well organized the changes are. I can pretty much follow his actions from readings his commit messages and looking at the grouped changes. :clap: :clap: :clap: 

Now on the broken unit tests. I got two broken tests:

```
# platform linux -- Python 3.7.0, pytest-3.8.0, py-1.6.0, pluggy-0.7.1
...
lib/cylc/tests/test_loggingutil.py::TestLoggingutil::test_value_error_raises_system_exit FAILED                                                                                                             [ 22%]
...
lib/cylc/tests/test_scheduler.py::TestScheduler::test_ioerror_is_ignored FAILED                                                                                                                             [ 31%]
...
================================================================================ 2 failed, 296 passed, 4 warnings in 18.36 seconds ================================================================================
```

First `test_scheduler.py::TestScheduler::test_ioerror_is_ignored`. This one was broken because Python 3 does a better job a type checking.

```
# This bugged line works in Python2
python -c 'import sys, os, mock; sys.path.append(os.path.join(mock.MagicMock(), "123"))'
# Same line fails in Python 3 (thankfully)
python3 -c 'import sys, os; from unittest import mock; sys.path.append(os.path.join(mock.MagicMock(), "123"))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.6/posixpath.py", line 80, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not MagicMock
```

FWIW, the code in CPython Python that checks it seems to be [this one](https://github.com/python/cpython/blob/9c3f284de598550be6687964c23fd7599e53b20e/Lib/os.py#L1033).

And `test_loggingutil.py::TestLoggingutil::test_value_error_raises_system_exit`. This one also had something that now is illegal in Python 3, comparing random objects with integers (again, related to more consistency with types, which is good).

```
# We had this in py2 before working
python -c 'import mock; mock.MagicMock() > 1'
# And now this one fails
python3 -c 'from unittest import mock; mock.MagicMock() > 1'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: '>' not supported between instances of 'MagicMock' and 'int'
```

@oliver-sanders sending a pull request showing what needs to be changed to that all unit tests pass. That doesn't need to be merged. You can pick up the changes and put them in a nice commit (or edit previous commit, etc) :-) this way I shouldn't mess with your current nice organisation.

```
kinow@kinow-VirtualBox:~/Development/python/workspace/cylc$ pytest -x 
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.7.0, pytest-3.8.0, py-1.6.0, pluggy-0.7.1 -- /home/kinow/Development/python/anaconda3/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/kinow/Development/python/workspace/cylc/.hypothesis/examples')
rootdir: /home/kinow/Development/python/workspace/cylc, inifile: pytest.ini
plugins: remotedata-0.3.0, openfiles-0.3.0, env-0.6.2, doctestplus-0.1.3, cov-2.6.0, arraydiff-0.2, hypothesis-3.83.1
collected 298 items                                                                                                                                                                                               

lib/cylc/conditional_simplifier.py::cylc.conditional_simplifier.ConditionalSimplifier.listify PASSED                                                                                                        [  0%]
...
===================================================================================== 298 passed, 4 warnings in 16.32 seconds =====================================================================================
```

Cheers
Bruno

oliver-sanders commented:
>> * Scan suites in parallel

> Or use asyncio?

@hjoliver Yes, I was looking at doing this yesterday, probably the best way forward.

Something that occurred to me is that unless we are requesting suite `[meta]` items or state totals we don't actually need to bother the suite to get the required information as it is stored in the contact file. Would this be a good idea?

* Scan results would appear faster
* Avoids unnecessary requests to suites
* Means that we don't actually check suites are running but just report the state of the filesystem

oliver-sanders commented:
> From a quick look, I guess you mean client UUID, not server UUID?

Yes, sorry I mean the client UUID. The ZMQ client-server I've implemented does not register clients or assign them a UUID.

wxtim commented on bin/cylc-profile-battery:
```diff
@@ -944,8 +946,8 @@ def main():
     if not prof.IS_GIT_REPO:
-        print >> sys.stderr, ('ERROR: profiling requires cylc to be a git '
-                              'repository.')
+        print(('ERROR: profiling requires cylc to be a git '
```
more elegant?
```python
        print('ERROR: profiling requires cylc to be a git repository',
              file=sys.stderr)
```

wxtim commented on bin/cylc-submit:
```diff
@@ -137,10 +137,11 @@ def main():
         for itask in itasks:
             if itask.local_job_file_path:
-                print('JOB SCRIPT=%s' % itask.local_job_file_path)
+                print(('JOB SCRIPT=%s' % itask.local_job_file_path))
```
Sorry, more unecessary brackets. :octocat:

wxtim commented on bin/cylc-validate:
```diff
@@ -129,20 +129,21 @@ def main():
             err = str(exc)
             if '@' in err:
-                print >> sys.stderr, (
+                print((
```
elegance suggestions only
```python
                print(("ERROR, %s: xtriggers can't be in conditional"
                       " expressions: %s" % (name, err)),
                      file=sys.stderr)

                print((f"ERROR, {name}: xtriggers can't be in conditional"
                       f" expressions: {err}"),
                      file=sys.stderr)
```

wxtim commented on bin/cylc-view:
```diff
@@ -172,13 +172,15 @@ def main():
+            'WARNING: YOU HAVE EDITED A TEMPORARY READ-ONLY SUITE COPY:'),
+            file=sys.stderr)
+        print(viewfile.name, file=sys.stderr)
+        print((
```
extra Brackets

wxtim commented on bin/cylc-view:
```diff
@@ -172,13 +172,15 @@ def main():
-            'In future use \'cylc [prep] edit\' to edit a suite.')
-        print
+        print()
+        print((
```
brackets

wxtim commented on tests/validate/45-jinja2-type-error.t:
```diff
@@ -51,7 +51,7 @@ run_fail "${TEST_NAME}" cylc validate 'suite.rc'
 Jinja2Error:
   File "<template>", line 2, in top-level template code
-ValueError: need more than 2 values to unpack
+ValueError: not enough values to unpack (expected 3, got 2)
```
:+1:

wxtim commented on lib/cylc/batch_sys_handlers/loadleveler.py:
```diff
@@ -55,7 +55,7 @@ def format_directives(self, job_conf):
                 job_conf["execution_time_limit"] + 60,
                 job_conf["execution_time_limit"])
-        for key, value in job_conf["directives"].items():
+        for key, value in list(job_conf["directives"].items()):
```
Is this list( statement really needed?

oliver-sanders replied:
It might be, there appears to be a circular nature to the `directives` variable, for safety I left the `list` in:

```python
directives = job_conf["directives"].__class__()

for key, value in list(job_conf["directives"].items():
    directives[key] = value
```

wxtim replied:
Fair Play.

wxtim commented on lib/cylc/broadcast_mgr.py:
```diff
@@ -87,7 +87,7 @@ def clear_broadcast(
                     while stuff_stack:
                         keys, stuff = stuff_stack.pop()
-                        for key, value in stuff.items():
+                        for key, value in list(stuff.items()):
```
Surplus list?

wxtim commented on tests/lib/python/test_diffr.py:
```diff
@@ -0,0 +1,414 @@
+import json
+import os
```
Unused import

wxtim commented on tests/lib/python/test_diffr.py:
```diff
@@ -0,0 +1,414 @@
+
+
+class TestDiff(unittest.TestCase):
+
```
Probably ought to have a test to ensure that sensible error is raised if int, str, float or any other non iterable is fed to it.

wxtim commented on tests/lib/python/diffr.py:
```diff
@@ -0,0 +1,300 @@
+        typ, meth = cls._diff_method(this, that)
+        if meth:
+            return typ, meth(this, that)
+        raise TypeError('%s Cannot compare %s and %s' % (
```
Should this raise be inside and else statement?

wxtim commented on tests/lib/python/diffr.py:
```diff
@@ -0,0 +1,300 @@
+"""Quick and simple JSON diff util.
+
+Examples:
+    >>> bool(Diff({}, {}))
```
I feel that `bool(Diff(A, B))` should return true if A and B are different, rather than the same!

oliver-sanders replied:
This is for consistency with CLI diff commands.

wxtim commented on tests/lib/python/diffr.py:
```diff
@@ -0,0 +1,300 @@
+    >>> list(diff.modified())
+    [(['b'], 2, 3)]
+
+    >>> diff = Diff({'x': {'y': {'z': 1}}}, {'x': {'y': {'z': 2}}})
```
Might you consider re-arranging this for ease of reading?
```python
diff = Diff({'x': {'y': {'z': 1}}},
            {'x': {'y': {'z': 2}}})
```

wxtim commented on tests/lib/python/diffr.py:
```diff
@@ -0,0 +1,300 @@
+        So unsurprisingly differencing nested lists is actually a little
+        tricky so this is the simple and crude method, just mark the items as
+        modified rather than going through the mess of working out how the
+        lists could fix together.
```
"fit together"?

hjoliver commented:
> Something that occurred to me is that unless we are requesting suite `[meta]` items or state totals we don't actually need to bother the suite to get the required information as it is stored in the contact file. Would this be a good idea?
>
>    * Scan results would appear faster
>    * Avoids unnecessary requests to suites
>    * Means that we don't actually check suites are running but just report the state of the filesystem

I suppose this would be fine, as we have to read the contact file anyway to find the suite location, right (for the users own suites)? And if the default is not to ask for the extra information, that would reduce load generally.  So I vote yes.

wxtim commented:
Suggested changes - my copy of cylc
test-battery tests/cli/01-help.t -v
was failing because
cylc profile-battery --help
failed with Python syntax errors.

https://github.com/oliver-sanders/cylc/pull/7

matthewrmshin commented:
As discussed off line. On client UUID. It is used for session management for the GUI, and has some minor roles for commands that send multiple requests to the suite server, e.g. `cylc monitor` and `cylc trigger --edit`. Otherwise, the only purpose is logging/accounting. As we currently have no capability to analyse suite logs, so client UUID can be considered redundant - now that the GUI is removed.

oliver-sanders commented:
#### OK, this should be ready for an in-depth review.

Note at present I'm unable to get the required Python3 environment onto any of our remote hosts so please run remote tests!

#### Questions to reviewers:
1. The client UUID has been functionally removed (though the `--print-uuid`, `--set-uuid` interface is still present). If we are happy with the removal I'll go through and tidy up.
2. Client requests used to be logged with some metadata (user, host and issuing program) do we want to keep this? It involves sending metadata with each request (see [patched tests](https://github.com/cylc/cylc/blob/85909b10c59ab2bfeb0ae7c04c363a708fc090f6/tests/authentication/05-full-control.t#L169))
3. Are we happy for cylc scan to scape contact files rather than connecting to suites where possible?
    * This would mean we don't actually check the suite is running, the only place this can cause problems is where suites have been hard killed.
    * The new `cylc scan` is running quickly for me, do we need to worry about this anyway?

#### Main areas to review:

* The new ZMQ interface
* The `cylc scan` changes

matthewrmshin commented:
Client UUID: remove for now.

Client meta data: good to keep, for future accounting trail analysis. Obviously we cannot trust it fully.

Cylc scan: happy either way. Perhaps use quick mode by default (i.e. contact files only), but connect when given an option? Or vice versa. It may appear fast now, but when you have real users, everything will slow down.

kinow commented on lib/cylc/task_remote_mgr.py:
```diff
@@ -272,7 +271,7 @@ def remote_tidy(self):
+                out, err = (f.decode() for f in proc.communicate())
                 if proc.wait():
                     LOG.warning(TaskRemoteMgmtError(
                         TaskRemoteMgmtError.MSG_TIDY,
```
Hummm, it appears to have a but on line 279. Not touched in this pull request, but I wonder if we can address it here... it's using `proc.ret_code`, but I just had a look at `subprocess.py` in my Anaconda Python installation, and it has only `.returncode`. Line 291 has actually a very similar code, but using `returncode`. If out of scope, I can create a separate issue for this one.

oliver-sanders replied:
Might as well fix it here.

oliver-sanders commented on lib/cylc/network/client.py:
```diff
@@ -0,0 +1,273 @@
+    """
+
+    DEFAULT_TIMEOUT = 1.  # 1 second
+    DEFAULT_TIMEOUT = 5.  # 5 second
```
A debug line, good spot.

oliver-sanders commented on bin/cylc-ping:
```diff
@@ -42,14 +42,13 @@ def main():
-    pclient.get_info('ping_suite')  # (no need to check the result)
+    pclient('ping_suite')  # (no need to check the result)
     if cylc.flags.verbose:
         sys.stdout.write("Running on %s:%s\n" % (pclient.host, pclient.port))
```
Sorted!

oliver-sanders commented on lib/cylc/version.py2:
```diff
@@ -0,0 +1,58 @@
+UNKNOWN = "UNKNOWN"
+
+
+def _get_cylc_version():
```
I have no idea why this exists, removed!

oliver-sanders commented on lib/cylc/profiling/profile.py:
```diff
@@ -292,7 +291,7 @@ def profile(schedule):
     results = {}
     success = True
-    for version_id, experiments in sorted(schedule.iteritems()):
+    for version_id, experiments in sorted(schedule.items()):
```
Your IDE parses docstrings, cool. Out-of-date docs are pretty darned bad, well spotted!

oliver-sanders commented on lib/cylc/scheduler.py:
```diff
@@ -56,9 +54,7 @@
-from cylc.suite_status import (
-    KEY_DESCRIPTION, KEY_GROUP, KEY_META, KEY_NAME, KEY_OWNER, KEY_STATES,
-    KEY_TASKS_BY_STATE, KEY_TITLE, KEY_UPDATE_TIME, KEY_VERSION)
+from cylc.suite_status import KEY_NAME, KEY_OWNER, KEY_VERSION
```
I'm getting the same message with my linter but it looks like it is used on line 953!

kinow replied:
I think the `KEY_NAME` used in [line 953](https://github.com/oliver-sanders/cylc/blob/cac70d18ff6da035cf59f2c12d83991ba9f7be2d/lib/cylc/scheduler.py#L953) is actually the property of the [suite_srv_files_mgr/SuiteSrvFilesManager class](https://github.com/oliver-sanders/cylc/blob/cac70d18ff6da035cf59f2c12d83991ba9f7be2d/lib/cylc/suite_srv_files_mgr.py#L55) ? So that import is still not used?

oliver-sanders commented:
Thanks for the thorough reviewing Bruno, some nice spots!

> Some TODO/NOTE markers were added

I was thinking of bumping these to GH issues once this PR is in.

oliver-sanders commented:
I appear to have broken `tests/restart/00-pre-initial.t`, will fix tomorrow, otherwise all OK.

wxtim commented:
> Thanks for the thorough reviewing Bruno, some nice spots!
> 
> > Some TODO/NOTE markers were added
> 
> I was thinking of bumping these to GH issues once this PR is in.

I've done that with most of the outstanding issues on Rose. These PRs are becoming worryingly large.

wxtim commented:
## Files Reviewed

- [ ] coverage.rc
- [ ] .gitignore
- [ ]

wxtim commented on .travis/coverage-data.sh:
```diff
@@ -16,6 +16,10 @@
+# call below tries to define the data_file, and sources locations for Travis.
+sed -e "s|data_file=.coverage|data_file=${TRAVIS_BUILD_DIR}/.coverage|g; s|./bin|${TRAVIS_BUILD_DIR}/bin|g; s|./lib|${TRAVIS_BUILD_DIR}/lib|g" .coveragerc > /tmp/.coveragerc
+# And some tests fail if we touch files in the git working directory, due
+# to Cylc's version appearing with the "dirty" suffix. To avoid this, we
```
Sentence truncated. The suspense is killing me.

wxtim commented on tests/lib/python/diffr.py:
```diff
@@ -0,0 +1,297 @@
+"""Quick and simple JSON diff util.
```
Copyright Header?

oliver-sanders replied:
I constantly forget this!

wxtim commented on tests/lib/python/test_diffr.py:
```diff
@@ -0,0 +1,414 @@
+"""Test the diff.py interface."""
```
Copyright bits

wxtim commented on tests/lib/python/graph.py:
```diff
@@ -0,0 +1,104 @@
+"""Standin for the text functions of the old cylc graph utility."""
```
Copyright stuff. :(

wxtim commented on tests/lib/python/test_diffr.py:
```diff
@@ -0,0 +1,414 @@
+        return [proc.returncode, stdout.decode(), stderr.decode()]
+
+    def test_dict_no_diff(self):
+        for this in [
```
I think I'd prefer `test_case`, `case` or `test_input` to `this` and `that`.

matthewrmshin commented:
I'm hoping to find time to go through this once more tomorrow. (Hopefully, I am not going to have another bug-of-the-day to deal with!)

Tests are OK (or are OK on retry) in my environment. This is in no way the final PR of the workflow service before Cylc 8 release, so I'm keen to get this merged as soon as possible to facilitate further improvements and other work in the pipeline.

hjoliver commented:
When this gets merged (soon as possible) I'll add a warning to the repository description, README, and post a note to the mail forum.  (cylc master branch now Python 3, new network layer, WIP, no GUI).  Hash-tag kick-ass.

hjoliver commented:
@oliver-sanders - presumably this should close #1874 (in which case, can you add the trigger to the description above).  And add your "Further Work" items to a new issue(s)?

hjoliver commented:
Note also #2858

oliver-sanders commented:
> presumably this should close #1874

Oh yes! I've added in related issues, should close a few with this one.

matthewrmshin commented on .travis.yml:
```diff
@@ -25,63 +25,93 @@
+
+  # run unittests at each supported Python version
+  - <<: *unittest
+    python: "3.6"
```
Any reason to support Python 3.6 just in here?

oliver-sanders replied:
We aught to tests the oldest and newest supported Python versions. Running the whole test battery under both Python versions is probably a bit much so what I've started here is:

* Run all tests under the latest Python version
* Run quick tests under the oldest supported Python Version

matthewrmshin commented on bin/cylc-cat-state:
```diff
@@ -99,18 +95,37 @@ def _get_state_lines(suite):
+    yield r"time : %(time_str)s (%(time_since_epoch)s)" % state
+    yield r"initial cycle : %(initial_point)s" % state
+    yield r"final cycle : %(final_point)s" % state
+    yield pickle.dumps(state["broadcast_states"]).splitlines()
```
We should aim to remove usage of pickle. Time to break compat? A TODO here is also acceptable if we want to defer to follow-on PR.

oliver-sanders replied:
Easy enough to swap out `pickle` for `json`.

matthewrmshin commented on bin/cylc-check-software:
```diff
@@ -49,29 +49,20 @@ latter grouped as Python, TeX or 'other' (neither). 'opt_spec' item format:
 <'OTHER' TUPLE> = ([<BASE CMD(S)>], <VER OPT>, <REGEX>, <OUTFILE ARG>).
 """
-req_py_ver_range = (2, 6), (3,)
+req_py_ver_range = ((3,),)
```
3.7+?

matthewrmshin commented on bin/cylc-check-versions:
```diff
@@ -111,30 +111,32 @@ def main():
-        print "WARNING: failed to invoke cylc-%s on %d accounts:" % (
-            CYLC_VERSION, len(warn.keys()))
+        print("WARNING: failed to invoke cylc-%s on %d accounts:" % (
+            CYLC_VERSION, len(list(warn.keys()))))
```
Just `len(warn)`?

matthewrmshin commented on bin/cylc-client:
```diff
@@ -32,39 +32,35 @@ if '--use-ssh' in sys.argv[1:]:
+    try:
+        main()
+    except Exception as exc:
+        sys.exit(exc)
```
I think we now want the simpler style for all the other CLI - as users are much happier seeing Python tracebacks.

```python
if __name__ == '__main__':
    main()
```

oliver-sanders replied:
>  as users are much happier seeing Python tracebacks

**I wouldn't be so sure of that!**

Traceback scares a lot of users and conceals the information the user actually needs.

I'm in two minds over this one, Cylc's historic habit of concealing traceback has caused problems with cryptic error messages which fail to describe the issue.

A simple rule of thumb for traceback might go something like:

* If the user hits an "expected" error (e.g. typo) give them a nice informative one line error.
* If the user hits an unexpected issue traceback may be more appropriate.

Take the example where a user gets the command wrong. Which is best from a UX perspective?

```console
$ # catching the exception
$ cylc client generic ping -n
Request returned error: No method by the name "ping"

$ # raising the exception
$ cylc client generic ping -n
Traceback (most recent call last):
  File "/cylc/bin/cylc-client", line 63, in <module>
    main()
  File "/cylc/bin/cylc-client", line 58, in main
    pclient(func, kwargs, timeout=options.comms_timeout), indent=4
  File "/cylc/lib/cylc/network/client.py", line 194, in serial_request
    self.async_request(command, args, timeout))
  File "/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "/cylc/lib/cylc/network/client.py", line 190, in async_request
    raise ClientError(response['error'])
cylc.network.client.ClientError: Request returned error: No method by the name "ping"
```

Now imagine you have never written a line of Python in your life and don't understand the workings of Cylc.

How other tools handle the same scenario:
```console
$ ssh myhost lss
bash: lss: command not found

$ git statts
WARNING: You called a Git command named 'statts', which does not exist.

$ ls --reversse
ls: unrecognized option '--reversse'

$ cylc stop elephant
Contact info not found for suite "elephant", suite not running?
```

Two supplemental matters:
1. The cylc client commands are a bit unusual as in the event of a server side error we only really want the server side error not the full client stack.
2. On a sidenote in Python3 if you catch an exception and raise a new one in its place you will get both exceptions in your traceback, the error the user needs to see might not be at the bottom of the trace in this example.

What should we do to make the interface informative and simple?

* Create an exception type for expected errors
* Always output the full traceback then underneath output `str(exc)` in bold to make the error clear.
* ?

matthewrmshin replied:
If possible, all known errors should subclass `CylcError` or something like that - so we only catch `CylcError` and leave all other Exception un-caught.

matthewrmshin replied:
(May be another TODO.)

oliver-sanders replied:
> If possible, all known errors should subclass CylcError

Probably the most sensible approach, we could stick a decorator on the `main()` methods to handle this.

For the moment I've removed the try/except, will raise a follow-on issue.

matthewrmshin commented on bin/cylc-cycle-point:
```diff
@@ -252,15 +252,15 @@ def main():
+            print(dump_string)
 
 
 if __name__ == "__main__":
```
Need to figure out how to make this a simple wrapper of `isodatetime`. @wxtim is on the case for `rose date`.

matthewrmshin commented on bin/cylc-show:
```diff
@@ -54,76 +54,73 @@ def main():
 
     if options.json:
-        print json.dumps(json_filter, indent=4)
+        print(json.dumps(json_filter, indent=4))
```
Not a porting issue, but we should remember to use `json.dump` (without the final `s`) in the future.

oliver-sanders replied:
Are you proposing:

```suggestion
        json.dump(json_filter, sys.stdout, indent=4))
```

matthewrmshin commented on doc/src/appendices/suiterc-config-ref.rst:
```diff
@@ -956,8 +953,8 @@ Allows tasks to spawn out to ``max active cycle points``
 *Important*: This should be used with care given the potential impact of
-additional task proxies both in terms of memory and cpu for the cylc daemon as
-well as overheads in rendering all the additional tasks in gcylc. Also, use
+additional task proxies in terms of memory and cpu for the cylc daemon.
```
Not a porting issue, but might as well update *cylc daemon* to *suite server program* or *workflow server program*.

wxtim replied:
Art thou requesting these changes here? Or do you want a new ticket?

matthewrmshin replied:
Either way. But given that the line is already touched, it is probably more convenient to incorporate.

matthewrmshin commented on doc/src/conf.py:
```diff
@@ -148,9 +148,9 @@
-   u'2008-2019 NIWA & British Crown (Met Office) & Contributors',
-  'cylc', 'The Cylc Suite Engine', 'Miscellaneous'),
+    ('index', 'cylc', 'Cylc Documentation',
+     '2008-2019 NIWA & British Crown (Met Office) & Contributors',
```
Not a porting issue, but it looks like we have a lot of repeated strings here.

matthewrmshin commented on doc/src/installation.rst:
```diff
@@ -10,85 +10,55 @@ including Apple OS X, but they are not officially tested and supported.
+Requirements:
 
-For Cylc's HTTPS communications layer:
+- Python 3.6+
```
Is this what we are going for?

oliver-sanders replied:
It's not what we're going for but I think it's what currently works!

matthewrmshin commented on doc/src/installation.rst:
```diff
@@ -10,85 +10,55 @@ including Apple OS X, but they are not officially tested and supported.
-The following packages are necessary for running all the tests in Cylc:
+The following packages are necessary for running tests in Cylc:
 
 - `mock <https://mock.readthedocs.io>`_
```
Is `mock` now part of standard library?

kinow replied:
Pre python 3 it was a standalone library. Post Python 3 I believe it was incorporated in `unittest` (not sure if 3.+ or if it was 3.3, 3.4, etc).

matthewrmshin replied:
Its doc says new in 3.3. If we are going for 3.6+ or 3.7+, we should no longer have to worry about it being an external dependency. (I don't think anyone would use 3.3 any more!? Even our brand new ancient RHEL7 desktop is on 3.4!)

matthewrmshin commented on lib/cylc/network/client.py:
```diff
@@ -0,0 +1,296 @@
+        if host and host.split('.')[0] == 'localhost':
+            host = get_host()
+        elif host and '.' not in host:  # Not IP and no domain
+            host = get_fqdn_by_host(host)
```
Should use `cylc.hostuserutil` module to these stuffs?

oliver-sanders replied:
Could move it there.

oliver-sanders replied:
Can do.

matthewrmshin commented on lib/cylc/network/scan.py:
```diff
@@ -0,0 +1,257 @@
+        awaiting.append(task)
+
+    index = 0
+    buff = []
```
Would it be better to use a `set` or `dict` here? Can we rename `buff` to `tasks` or `completed_tasks` here?

oliver-sanders replied:
* Will rename.
* `buff` needs to maintain order so `set` wouldn't work.
* items don't have associated keys so `dict` wouldn't be the first choice.

I must be missing something here, what's your logic?

matthewrmshin replied:
Since the logic appears to be doing a lot of `.extend` and `.remove`. More natural to use `dict`?

matthewrmshin commented on lib/cylc/network/scan.py:
```diff
@@ -0,0 +1,257 @@
+    return (cres['owner'], cres['name'])
+
+
+def get_scan_items_from_fs(owner_pattern=None, reg_pattern=None, updater=None):
```
Can probably remove `updater`. It was mainly so that `gscan` can be Ctrl-C.

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        """Register all exposed methods."""
+        self.endpoints = {name: obj
+                          for name, obj in self.__class__.__dict__.items()
+                          if hasattr(obj, 'exposed')}
```
Syntax a bit ugly.

oliver-sanders replied:
Would you rather:

```python
        self.endpoints = {
            name: obj
            for name, obj in self.__class__.__dict__.items()
            if hasattr(obj, 'exposed')
        }
```

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+                self.port = port
+                break
+        else:
+            raise Exception('No room at the inn, all ports occupied.')
```
IOError?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        LOG.debug('stopping zmq server...')
+        self.queue.put('STOP')
+        self.thread.join()  # wait for the listener to return
+        LOG.debug('...stopped')
```
Is this called on KeyboardInterrupt?

oliver-sanders replied:
Yep.

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+            # process
+            try:
+                message = self.decode(msg, self.secret())
+                LOG.debug('zmq:recv %s', message)
```
LOG should live under `else:`?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+            response = method(**args)
+        except Exception as exc:
+            # includes incorrect arguments (TypeError)
+            LOG.error(exc)  # note the error server side
```
LOG.exception?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+
+            usr_priv_level = self.get_priv_level(user)
+            if usr_priv_level < req_priv_level:
+                LOG.info(
```
LOG.warn?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+            ungroup_nodes=ungroup_nodes,
+            ungroup_recursive=ungroup_recursive,
+            group_all=group_all,
+            ungroup_all=ungroup_all)
```
We really need to do something about this. Not part of this PR though!

oliver-sanders replied:
Yep it's pretty horrible.

oliver-sanders commented on lib/cylc/log_diagnosis.py:
```diff
@@ -29,7 +29,7 @@ def __init__(self, msg):
 
     def __str__(self):
-        return repr(self.msg)
+        return self.msg
```
Yep, actually we can just do

```python
class LogAnalyserError(Exception): pass
```

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+    def get_task_requisites(self, items=None, list_prereqs=False):
+        """Return prerequisites of a task."""
+        if not isinstance(items, list):
+            items = [items]
```
In the old REST API, whether items is a list or not may be ambiguous. I cannot see why this will still be the case in the new API?

oliver-sanders replied:
Will strip these typing statements.

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        items is a list of identifiers for matching task proxies.
+        """
+        if not isinstance(items, list):
+            items = [items]
```
Unnecessary in new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        if not isinstance(items, list):
+            items = [items]
+        if stop_point_string == "None":
+            stop_point_string = None
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        items is a list of identifiers for matching task proxies.
+        """
+        if not isinstance(items, list):
+            items = [items]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        items is a list of identifiers for matching task proxies.
+        """
+        if items is not None and not isinstance(items, list):
+            items = [items]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        items is a list of identifiers for matching task proxies.
+        """
+        if not isinstance(items, list):
+            items = [items]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        items is a list of identifiers for matching task proxies.
+        """
+        if not isinstance(items, list):
+            items = [items]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        if not isinstance(items, list):
+            items = [items]
+        if outputs and not isinstance(outputs, list):
+            outputs = [outputs]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        items is a list of identifiers for matching task proxies.
+        """
+        if not isinstance(items, list):
+            items = [items]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        items[0] is the name of the checkpoint.
+        """
+        if not isinstance(items, list):
+            items = [items]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/network/server.py:
```diff
@@ -0,0 +1,673 @@
+        """
+        if not isinstance(items, list):
+            items = [items]
+        items = [str(item) for item in items]
```
Unnecessary in the new API?

matthewrmshin commented on lib/cylc/suite_srv_files_mgr.py:
```diff
@@ -506,98 +497,20 @@ def register(self, reg=None, source=None, redirect=False):
             os.symlink(source_str, target)
 
-        print('REGISTERED %s -> %s' % (reg, source))
+        print(('REGISTERED %s -> %s' % (reg, source)))
```
Too many brackets?

matthewrmshin commented on lib/cylc/tests/test_task_outputs.py:
```diff
@@ -0,0 +1,31 @@
+import random
```
Need standard header.

matthewrmshin commented on lib/parsec/fileparse.py:
```diff
@@ -289,7 +289,7 @@ def read_and_proc(fpath, template_vars=None, viewcfg=None, asedit=False):
             try:
                 flines = jinja2process(flines, fdir, template_vars)
-            except (StandardError, TemplateError, UndefinedError) as exc:
+            except (Exception, TemplateError, UndefinedError) as exc:
```
Does Exception not cover everything?

oliver-sanders replied:
I'm not sure what the original intention was here!

```suggestion
            except Exceptionas exc:
```

???

matthewrmshin replied:
Yes.

matthewrmshin commented on lib/parsec/fileparse.py:
```diff
@@ -302,7 +302,7 @@ def read_and_proc(fpath, template_vars=None, viewcfg=None, asedit=False):
                 if hasattr(exc, 'lineno'):
                     lineno = exc.lineno
-                elif (isinstance(exc, StandardError) or
+                elif (isinstance(exc, Exception) or
```
Is this always True?

oliver-sanders replied:
I'm not sure what the `StandardError` logic was meant to achive.

matthewrmshin replied:
We probably don't need the `or ...` bit here?

matthewrmshin commented on tests/restart/00-pre-initial.t:
```diff
@@ -22,6 +22,9 @@ install_suite "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
 suite_run_ok "${TEST_NAME_BASE}-run" cylc run --debug --no-detach "${SUITE_NAME}"
+while cylc ping "${SUITE_NAME}" 2>/dev/null; do
+    sleep 1  # wait for suite to shut down
+done
```
Should `--no-detach` already handle this?

matthewrmshin commented on tests/cylc-trigger/07-edit-run-abort/suite.rc:
```diff
@@ -26,7 +26,7 @@ sleep 2
 
 my-suite-state-summary-test \
-    < <(cylc client -n 'get_suite_state_summary' "${CYLC_SUITE_NAME}")
+    <<<$(cylc client -n "${CYLC_SUITE_NAME}" 'get_suite_state_summary')
```
Why do we need to change the process substitution to a here string?

matthewrmshin commented on tests/cylc-show/prereqs-outputs/suite.rc:
```diff
@@ -49,11 +60,19 @@
+            cylc trigger ${CYLC_SUITE_NAME} baz.1
+            cylc suite-state "${CYLC_SUITE_NAME}" \
+                -t baz -p 1 -S start --interval=1
+        """
```
I am still none the wiser with this test!

oliver-sanders replied:
I'm not much better but it is at least stable.

matthewrmshin commented:
All minor comments.

matthewrmshin commented:
Codacy has a few legit unused imports, unused statements, etc. You may want to take note.

kinow commented:
I believe before or after this issue is merged, we can close https://github.com/cylc/cylc/issues/2647, as cylc_xdot.py is no more.

oliver-sanders commented on lib/cylc/cfgspec/globalcfg.py:
```diff
@@ -104,7 +103,7 @@
-        'method': [VDR.V_STRING, 'https', 'http'],
+        'method': [VDR.V_STRING, 'zmq'],
         'options': [VDR.V_STRING_LIST],
     },
```
delete, delete, delete

matthewrmshin replied:
Delete.

oliver-sanders replied:
On the task...

Have removed `KEY_COMMS_PROTOCOL`, however there is also `KEY_COMMS_PROTOCOL_2` which would appear to be related to task comms but I'm not convinced this key is used. Is is safe to remove?

matthewrmshin replied:
(That was probably for SSH+HTTP(S) so we should delete here. Ref #2975.)

matthewrmshin commented:
I have added some replies to the unresolved conversations. Address as necessary. This is good to go otherwise.

oliver-sanders commented:
* Addressed feedback on unrelated code.
* Removed `[communications]` section
* Removed HTTP(s) references from the documentation
* Kept SSH tunnelling references in documentation awaiting #2975

matthewrmshin commented:
@hjoliver 2 approvals. Do you want to take a look of this as well? Or are you happy for this to be merged?

(Note: subject to Travis CI returning a good status.)

hjoliver commented:
I'm happy to merge.  I'm (sadly) still not sufficiently au-fait with Python 3 differences to contribute much, and we don't need to fine-tune at this stage (cylc-8 is still almost a year away...).  ... Thanks all, brilliant 💥

hjoliver commented:
Travis CI passed this 😁

hjoliver commented:
Codacy issues are not important, can be addressed later.

dwsutherland commented:
Nice! :boom: 
No going back now!
